lib/dataset/bin.cpp

```diff
@@ -30,74 +31,186 @@ namespace scipp::dataset {
 
 namespace {
 
-template <class T, class Builder>
-auto setup_and_apply(const Variable &data, const Variable &indices,
-                     const Builder &builder) {
-  const auto dims = builder.dims();
-  // Setup offsets within output bins, for every input bin. If rebinning occurs
-  // along a dimension each output bin sees contributions from all input bins
-  // along that dim.
-  auto output_bin_sizes =
-      bin_detail::bin_sizes(indices, builder.offsets(), builder.nbin());
-  auto offsets = copy(output_bin_sizes);
-  fill_zeros(offsets);
-  // Not using cumsum along *all* dims, since some outer dims may be left
-  // untouched (no rebin).
-  std::vector<std::pair<Dim, scipp::index>> strategy;
-  for (const auto dim : data.dims())
-    if (dims.contains(dim))
-      strategy.emplace_back(dim, data.dims()[dim]);
-  // To avoid excessive memory consumption in intermediate results for
-  // `output_bin_sizes` (in the loop below, computing sums and cumsums) we need
-  // to ensure to handle the longest dimensions first,
-  std::sort(strategy.begin(), strategy.end(),
-            [](auto &&a, auto &&b) { return a.second > b.second; });
-  for (const auto &item : strategy) {
-    const auto dim = item.first;
+template <class T>
+Variable bins_from_sizes(T &&content, const Variable &bin_sizes) {
+  const auto end = cumsum(bin_sizes);
+  const auto buffer_dim = content.dims().inner();
+  return make_bins(zip(end - bin_sizes, end), buffer_dim, std::move(content));
+}
+
+template <class Builder> bool use_two_stage_remap(const Builder &bld) {
+  return bld.nbin().dims().empty() &&
+         bld.nbin().template value<scipp::index>() == bld.dims().volume() &&
+         // empirically determined crossover point (approx.)
+         bld.nbin().template value<scipp::index>() > 16 * 1024 &&
+         bld.offsets().dims().empty() &&
+         bld.offsets().template value<scipp::index>() == 0;
+}
+class Mapper {
+public:
+  template <class T> T do_bin(const Variable &data) {
+    const auto maybe_bin = [this](const auto &var) {
+      return is_bins(var) ? _do_bin(var) : copy(var);
+    };
+    if constexpr (std::is_same_v<T, Variable>)
+      return maybe_bin(data);
+    else
+      return dataset::transform(bins_view<T>(data), maybe_bin);
+  }
+
+  virtual Variable bin_sizes(
+      const std::optional<Dimensions> &dims_override = std::nullopt) const = 0;
+  virtual Variable _do_bin(const Variable &var) const = 0;
+};
+
+class SingleStageMapper : public Mapper {
+public:
+  SingleStageMapper(const Dimensions &dims, const Variable &indices,
+                    const Variable &output_bin_sizes)
+      : m_dims(dims), m_indices(indices), m_output_bin_sizes(output_bin_sizes) {
+    // Setup offsets within output bins, for every input bin. If rebinning
```

Comment:
The code in this constructor is essentially just moved from the old `setup_and_apply`.